src/systems/physics/Physics.cc

```diff
@@ -1549,43 +1679,50 @@ void PhysicsPrivate::UpdateSim(EntityComponentManager &_ecm)
           auto canonicalLink =
               _ecm.Component<components::CanonicalLink>(_entity);
 
-          // get the pose component of the parent model
-          const components::Pose *parentPose =
-              _ecm.Component<components::Pose>(_parent->Data());
-
           auto frameData = linkIt->second->FrameDataRelativeToWorld();
           const auto &worldPose = frameData.pose;
 
-          // if the parentPose is a nullptr, something is wrong with ECS
-          // creation
-          if (!parentPose)
-          {
-            ignerr << "The pose component of " << _parent->Data()
-                   << " could not be found. This should never happen!\n";
-            return true;
-          }
           if (canonicalLink)
           {
-            // This is the canonical link, update the model
-            // The Pose component, _pose, of this link is the initial
-            // transform of the link w.r.t its model. This component never
-            // changes because it's "fixed" to the model. Instead, we change
-            // the model's pose here. The physics engine gives us the pose of
-            // this link relative to world so to set the model's pose, we have
-            // to post-multiply it by the inverse of the initial transform of
-            // the link w.r.t to its model.
-            auto mutableParentPose =
-              _ecm.Component<components::Pose>(_parent->Data());
-            *(mutableParentPose) = components::Pose(_pose->Data().Inverse() +
-                                           math::eigen3::convert(worldPose));
+            // This is the canonical link, update the top level model.
```

Comment:
This might be problematic because nested models have their own canonical links in addition to the canonical link of the containing model. While iterating through the links, you might have multiple canonical links that all have the same top level model. I'm not sure what would be the best way to avoid this. It would be nice if we could tell which canonical link belongs to which model in the hierarchy. This would make it possible to change the pose of nested models using the transform tool, something that isn't available in [gazebo-classic](https://answers.gazebosim.org/question/25506/sdf-worldmodel-composition-across-files-is-impossible-groupingempty-containers-of-models/). But right now the `components::CanonicalLink` component doesn't give us enough information to make that association. Maybe we can add a pointer to the model entity in the component, but that will have to wait for a future version because it breaks ABI. 

For now, the only solution I can think of is to ensure that only one canonical link per model hierarchy exists by construction in SdfEntityCreator.

Reply:
sounds good. I added a helper function `SdfEntityCreator::CreateEntities(sdf::Model _model, bool _createCanonicalLink)` function that recusrively creates the model while keeping track of whether a canonical link has been created or not in a model tree.